nalgebra-sparse/src/cs.rs

```diff
@@ -543,3 +544,146 @@ pub fn convert_counts_to_offsets(counts: &mut [usize]) {
+        if !first_offset_ok || !last_offset_ok {
+            return Err(SparseFormatError::from_kind_and_msg(
+                SparseFormatErrorKind::InvalidStructure,
+                "First or last offset is incompatible with format.",
+            ));
+        }
+    }
+
+    let mut minor_index_permutation: Vec<usize> = Vec::new();
+
+    // Test that each lane has strictly monotonically increasing minor indices, i.e.
+    // minor indices within a lane are sorted, unique. Sort minor indices within a lane if needed.
+    // In addition, each minor index must be in bounds with respect to the minor dimension.
+    {
+        for lane_idx in 0..major_dim {
+            let range_start = major_offsets[lane_idx];
+            let range_end = major_offsets[lane_idx + 1];
+
+            // Test that major offsets are monotonically increasing
+            if range_start > range_end {
+                return Err(SparseFormatError::from_kind_and_msg(
+                    SparseFormatErrorKind::InvalidStructure,
+                    "Offsets are not monotonically increasing.",
+                ));
+            }
+
+            let minor_idx_in_lane = minor_indices.get(range_start..range_end).ok_or(
+                SparseFormatError::from_kind_and_msg(
+                    SparseFormatErrorKind::IndexOutOfBounds,
+                    "A major offset is out of bounds.",
+                ),
+            )?;
+
+            // We test for in-bounds, uniqueness and monotonicity at the same time
+            // to ensure that we only visit each minor index once
+            let mut iter = minor_idx_in_lane.iter();
+            let mut prev = None;
+            let mut nonmonotonic = false;
+
+            while let Some(next) = iter.next().copied() {
+                if next >= minor_dim {
+                    return Err(SparseFormatError::from_kind_and_msg(
+                        SparseFormatErrorKind::IndexOutOfBounds,
+                        "A minor index is out of bounds.",
+                    ));
+                }
+
+                if let Some(prev) = prev {
+                    if prev == next {
+                        return Err(SparseFormatError::from_kind_and_msg(
+                            SparseFormatErrorKind::DuplicateEntry,
+                            "Input data contains duplicate entries.",
+                        ));
+                    }
+                    if prev > next {
+                        nonmonotonic = true;
+                    }
+                }
+                prev = Some(next);
+            }
+
+            // sort if indices are nonmonotonic and sorting is expected
+            if nonmonotonic && sort {
+                minor_index_permutation.resize(range_end - range_start, 0);
+                for (index, value) in (range_start..range_end).enumerate() {
+                    minor_index_permutation[index] = value;
+                }
```

Comment:
Didn't we have some functions for applying permutations already? I think this whole part would be much easier to understand if we put the permutation stuff into separate functions. For example (not in any way tested):

```rust
// Set up *all* buffers up front (i.e. at the beginning of function, e.g. `Vec::new()`
let mut minor_idx_buffer = Vec::new();
let mut values_buffer = Vec::new();

// ...

compute_sort_permutation(&mut minor_index_permutation, &minor_idx_in_lane);
minor_idx_buffer.resize(minor_idx_in_lane.len());
permute(&minor_index_permutation, &minor_idx_in_lane, &mut minor_idx_buffer);
minor_idx_in_lane.copy_from_slice(&minor_idx_buffer);

if Some(values) = values {
   permute(&minor_index_permutation, &values, &mut values_buffer);
   values.clone_from_slice(&values_buffer);
}
```

What do you think?

Reply:
Doesn't have to look exactly like this, but the most important part is to remove the complete allocation of a buffer on every lane, so that we can amortize allocations over all lanes.

Reply:
Moving some logic from here to separate functions seems to work fine, I think it's a good idea.
I created _utils.rs_ for this purpose again.